src/py21cmfast/wrapper.py

```diff
@@ -3219,3 +3223,308 @@ def calibrate_photon_cons(
+                dtau_new = np.take(dtau_new, a, axis=0, mode="wrap")
+                dtau_new = np.take(dtau_new, a, axis=1, mode="wrap")
+                Tcmb_new = np.take(Tcmb_new, a, axis=0, mode="wrap")
+                Tcmb_new = np.take(Tcmb_new, a, axis=1, mode="wrap")
+            if rotation:
+                if k % kSZ_consts.HII_DIM == 0:
+                    tx = int(kSZ_consts.HII_DIM * random.random())
+                    ty = int(kSZ_consts.HII_DIM * random.random())
+                dtau_new = np.roll(
+                    dtau_new, -tx, 0
+                )  # shifting of tcmb so there is no object repetition
+                dtau_new = np.roll(dtau_new, -ty, 1)
+                Tcmb_new = np.roll(
+                    Tcmb_new, -tx, 0
+                )  # shifting of tcmb so there is no object repetition
+                Tcmb_new = np.roll(Tcmb_new, -ty, 1)
+            taue_arry += dtau_new  # tau_e updating
+            Tcmb += Tcmb_new * np.exp(
+                -taue_arry
+            )  # tcmb contribution with tau_e taken in account
+    mean_taue_fin = np.mean(taue_arry)
+    Tcmb = Tcmb - np.mean(Tcmb)
+    return Tcmb, mean_taue_fin
+
+
+class KszConstants:
+    """Constants used for kSZ calculation."""
+
+    def __init__(
+        self,
+        HII_DIM,
+        BOX_LEN,
+        hlittle,
+        OMb,
+        red_dist,
+        redshift_start,
+        DA_zstart,
+        Y_He,
+        Zreion_HeII
+        #    mean_taue_curr_z
+    ):
+        RHOb_cgs = (
+            3.0
+            * (hlittle * 3.2407e-18) ** 2
+            / (8.0 * np.pi * constants.G.cgs.value)
+            * OMb
+            / constants.m_p.cgs.value
+        )  # pcm^-3 at z=0
+        self.He_No = (
+            RHOb_cgs * self.Y_He / 4.0
+        )  # current helium number density estimate
+        self.N_b0 = RHOb_cgs * (
+            1 - 0.75 * self.Y_He
+        )  # present-day baryon num density, H + He
+        self.dR = BOX_LEN / HII_DIM
+        self.CMperMPC = constants.kpc.cgs.value * 1e3
+        self.A = self.N_b0 * constants.sigma_T.cgs.value * self.dR * self.CMperMPC
+        self.HII_DIM = HII_DIM
+        self.BOX_LEN = BOX_LEN
+        self.red_dist = red_dist
+        self.redshift_start = redshift_start
+        self.DA_zstart = DA_zstart
+        self.Zreion_HeII = Zreion_HeII
+
+
+def drdz(z, cosmo_params):
+    """Comiving distance (in cm) per unit redshift."""
+    return (1.0 + z) * constants.c.cgs.value * dtdz(z, cosmo_params)
+
+
+def dtdz(z, cosmo_params):
+    """The value of dt/dz at the redshift parameter z."""
+    OMm = cosmo_params.OMm
+    OMl = 1 - OMm
+    hlittle = cosmo_params.hlittle
+    Ho = hlittle * 3.2407e-18  # /* Hubble parameter[s^-1] at z=0 */
+
+    x = np.sqrt(OMl / OMm) * (1 + z) ** (-3.0 / 2.0)
+    dxdz = np.sqrt(OMl / OMm) * (1 + z) ** (-5.0 / 2.0) * (-3.0 / 2.0)
+    const1 = 2 * np.sqrt(1 + OMm / OMl) / (3.0 * Ho)
+    numer = dxdz * (1 + x * (x ** 2 + 1) ** (-0.5))
```

Comment:
not checking these functions, plz make sure they are correct :)

Reply:
As I've said in slack, I can probably erase this since it's only used for tau_e calculation.